### ISSUE TITLE
Discard duplicate jobs

### ISSUE THREAD

**Issue by matteeyah** [rails/solid_queue#176]
## Summary

The concurrency controls that currently exist allow blocking execution of "duplicate" jobs, or jobs with the same arguments. These jobs wait, then execute after the currently running job finishes. There's no way to discard all duplicate jobs and prevent them from running completely.

## Proposal

Add a way to discard jobs that get blocked by the concurrency controls.

**Comment by rosa** [comment]
@matteeyah, I've got this on my mind as well! In the end, I'm not sure there's a version of https://github.com/basecamp/solid_queue/issues/105 that I like or that it's useful for what we need, so I thought, in the meantime, to add another parameter to `limits_concurrency` to discard instead of blocking. Something like: 
```ruby
on_conflict: :discard
```
By default, it'd be: 
```ruby
on_conflict: :block
```
which would be the current behaviour.

**Comment by matteeyah** [comment]
That's exactly what I had in mind!

---

For reference - We're currently using https://github.com/veeqo/activejob-uniqueness. It serves a slightly different purpose, but it also checks for duplicate jobs.

It also has an `on_conflict` option - it supports logging a conflict, raising an error on a conflict and passing a proc to handle the conflict manually.

**Comment by nhorton** [comment]
Another vote for following the activejob-uniqueness conventions.

I would note that of the strategies offered by the gem, `until_and_while_executing` is the one we have found most-useful.

**Comment by nhorton** [comment]
Is there any more word on this? Or any openness to others trying to contribute?

**Comment by rosa** [comment]
Hey @nhorton, sorry for the delay! I'm back at this now and would like to get this feature ready for version 1.0. 

> of the strategies offered by the gem, `until_and_while_executing` is the one we have found most-useful.

I think that would be the only strategy supported because of the way concurrency controls work. Right now jobs don't unblock other jobs or release the semaphore until they complete, and for simplicity, it'd be much easier to keep it that way. 

> Or any openness to others trying to contribute?

Definitely! 🙌

**Comment by bilby91** [comment]
@rosa Is there an ETA for 1.0 ?

**Comment by rosa** [comment]
@bilby91, my hope is by the end of August but you know, there are always unexpected things 😅 😅

**Comment by bilby91** [comment]
@rosa Awesome! I look forward to testing this feature.

**Comment by rosa** [comment]
In the end I won't be able to get this one in for v1.0, but I hope to get it there shortly after. Sorry!

**Comment by wheee** [comment]
Hey @rosa , any chance that this will come about sometime in the near/medium future?  Would love to use this feature as well!

**Comment by joelzwarrington** [comment]
Hey @rosa! Any chance you can have a look at #523 sometime soon? Thanks :bow:

**Comment by rosa** [comment]
@joelzwarrington yes, sorry! I was out at a conference since Wednesday last week and came back yesterday, and then had to caught up on a bunch of other stuff at work.

**Comment by rosa** [comment]
This was finally done in https://github.com/rails/solid_queue/pull/594, and released in the last version, v.1.2.0.
